lib/Infer/Interpreter.cpp

```diff
@@ -0,0 +1,269 @@
+      return {Inst->Val};
+    case souper::Inst::UntypedConst:
+      return {Inst->Val};
+    case souper::Inst::Var:
+      llvm_unreachable("Should get value from cache without reaching here");
+    case souper::Inst::Add:
+      return {args[0].getValue() + args[1].getValue()};
+    case souper::Inst::Sub:
+      return {args[0].getValue() - args[1].getValue()};
+    case souper::Inst::Mul:
+      return {args[0].getValue() * args[1].getValue()};
+    case souper::Inst::UDiv:
+      return {args[0].getValue().udiv(args[1].getValue())};
+    case souper::Inst::SDiv:
+      return {args[0].getValue().sdiv(args[1].getValue())};
+    case souper::Inst::URem:
+      return {args[0].getValue().urem(args[1].getValue())};
+    case souper::Inst::SRem:
+      return {args[0].getValue().srem(args[1].getValue())};
+    case souper::Inst::And:
+      return {args[0].getValue() & args[1].getValue()};
+    case souper::Inst::Or:
+      return {args[0].getValue() | args[1].getValue()};
+    case souper::Inst::Xor:
+      return {args[0].getValue() ^ args[1].getValue()};
+    case souper::Inst::Shl:
+      return {args[0].getValue() << args[1].getValue()};
+    case souper::Inst::LShr:
+      return {args[0].getValue().lshr(args[1].getValue())};
+    case souper::Inst::AShr:
+      return {args[0].getValue().ashr(args[1].getValue())};
+    // TODO: Handle NSW/NUW/NW variants of instructions
+    case souper::Inst::Select: {
+      if (!args[0].hasValue()) {
+        return args[0];
+      } else {
+        bool cond = args[0].getValue().getBoolValue();
+        return cond ? args[1] : args[2];
+      }
+    }
+    case souper::Inst::ZExt:
+      return {args[0].getValue().zext(Inst->Width)};
+    case souper::Inst::SExt:
+      return {args[0].getValue().sext(Inst->Width)};
+    case souper::Inst::Trunc:
+      return {args[0].getValue().trunc(Inst->Width)};
+    case souper::Inst::Eq:
+      return {{1, args[0].getValue() == args[1].getValue()}};
+    case souper::Inst::Ne:
+      return {{1, args[0].getValue() != args[1].getValue()}};
+    case souper::Inst::Ult:
+      return {{1, args[0].getValue().ult(args[1].getValue())}};
+    case souper::Inst::Slt:
+      return {{1, args[0].getValue().slt(args[1].getValue())}};
+    case souper::Inst::Ule:
+      return {{1, args[0].getValue().ule(args[1].getValue())}};
+    case souper::Inst::Sle:
+      return {{1, args[0].getValue().sle(args[1].getValue())}};
+    case souper::Inst::CtPop:
+      return {llvm::APInt(Inst->Width, args[0].getValue().countPopulation())};
+    case souper::Inst::Ctlz:
+      return {llvm::APInt(Inst->Width,
+                          args[0].getValue().countLeadingZeros())};
+    case souper::Inst::Cttz:
+      return {llvm::APInt(Inst->Width,
+                          args[0].getValue().countTrailingZeros())};
+    case souper::Inst::BSwap:
+      return {args[0].getValue().byteSwap()};
+    default:
+      return EvalValue(); // Indicates an 'unavailable' value
```

Comment:
please change mentions of unavailable to unimplemented or whatever is correct now